platform/android/src/jni.cpp

```diff
@@ -373,6 +435,77 @@ std::pair<mbgl::AnnotationSegment, mbgl::ShapeAnnotation::Properties> annotation
     return std::make_pair(segment, shapeProperties);
 }
 
+/*
+ * Adapted from https://github.com/dropbox/djinni/blob/master/support-lib/jni/Marshal.hpp
+ */
+
+void jniExceptionCheck(JNIEnv * env) {
+    if (!env) {
+        mbgl::Log::Error(mbgl::Event::JNI, "Empty environment.");
+        abort();
+    }
+
+    const jthrowable e(env->ExceptionOccurred());
+    if (e) {
+        env->ExceptionClear();
+    }
+}
+
+static std::vector<uint8_t> metadata_from_java(JNIEnv* jniEnv, jbyteArray j) {
+    assert(j != nullptr);
+
+    std::vector<uint8_t> ret;
+    jsize length = jniEnv->GetArrayLength(j);
+    jniExceptionCheck(jniEnv);
+
+    if (!length) {
+        mbgl::Log::Error(mbgl::Event::JNI, "jbyteArray was empty.");
+        return ret;
+    }
+
+    {
+        auto deleter = [jniEnv, j] (void* c) {
+            if (c) {
+                jniEnv->ReleasePrimitiveArrayCritical(j, c, JNI_ABORT);
+            }
+        };
+
+        std::unique_ptr<uint8_t, decltype(deleter)> ptr(
+            reinterpret_cast<uint8_t*>(jniEnv->GetPrimitiveArrayCritical(j, nullptr)),
+            deleter
+        );
+
+        if (ptr) {
+            // Construct and then move-assign. This copies the elements only once,
+            // and avoids having to initialize before filling (as with resize())
```

Comment:
I have simplified this code in https://github.com/mapbox/mapbox-gl-native/commit/7f7f492219c29e800d190d54aa656c984a917538 and I found the cause of the crash, it was actually in a different spot where I was overwriting the actual class, not its `byte[]` object.